buildSrc/src/main/groovy/Docker.groovy

```diff
@@ -337,35 +364,34 @@ class Docker {
                 }
             }
 
-            // Handle copying failure. This is now distinct from the "actual" Sync task that depends directly
-            // on the CombinedDockerRunTask.
-            def syncAfterFail = project.tasks.register("${taskName}SyncAfterFail", Sync) { sync ->
+            // Specify that makeImage is finalized by buildAndRun - that is, in this configuration buildAndRun
+            // must run after makeImage finishes
+            makeImage.configure {it ->
+                it.finalizedBy(buildAndRun)
+            }
+
+            // Handle copying output from the docker task to the user-controlled location
+            def syncOutput = project.tasks.register(taskName, Sync) { sync ->
                 sync.with {
+                    dependsOn(makeImage)
                     // run the provided closure first
                     cfg.copyOut.execute(sync)
 
                     // then set the from location
                     from dockerCopyLocation
 
-                    onlyIf { buildAndRun.get().state.failure != null }
+                    doLast {
+                        if (buildAndRun.get().state.failure != null) {
+                            throw new GradleException('Docker task failed, see earlier task failures for details')
```

Comment:
I might add a comment this isn't an illegal state, we just want to preserve / propagate the failure, correct? I wonder if we should have a separate task that propagates this failure, instead of attaching it to the Sync task? (which make have specific semantics attached to sync-related failures.)

Reply:
Correct - not an illegal state (and the choice of GradleException is I thought intended to convey that), but just a failure. If this task did not throw an exception (or otherwise fail), it wouldn't be clear to other tasks that they should/shouldn't continue (e.g. "if it failed, dont continue to build downstream artifacts" vs "if it failed, make sure we dump the server's logs").

Reply:
Note also that a new task with actions (i.e. "fail if that other thing failed") but no inputs or outputs can never be up to date. I'm not aware of Sync specific errors, but can add a comment here to call out what is going on, besides the wall of text written at the top of this section.